### PR TITLE
Task/sdk 3236/migrate to work manager

### PR DESCRIPTION
**Pull request by Anush-Shand**
https://wizrocket.atlassian.net/browse/SDK-3236

### CODE REVIEW

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java [hunk 1]
```diff
@@ -518,70 +478,84 @@ public Void call() {
 
     @SuppressLint("MissingPermission")
     @RequiresApi(api = VERSION_CODES.LOLLIPOP)
-    private void createOrResetJobScheduler(Context context) {
-
+    private void stopJobScheduler(Context context) {
         int existingJobId = StorageHelper.getInt(context, Constants.PF_JOB_ID, -1);
-        JobScheduler jobScheduler = (JobScheduler) context.getSystemService(JOB_SCHEDULER_SERVICE);
+        if (existingJobId >= 0) {
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:483]: != 1

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java [hunk 2]
```diff
@@ -518,70 +478,84 @@ public Void call() {
...
-            int jobid = config.getAccountId().hashCode();
-            JobInfo.Builder builder = new JobInfo.Builder(jobid, componentName);
-            builder.setRequiredNetworkType(JobInfo.NETWORK_TYPE_ANY);
-            builder.setRequiresCharging(false);
-
-            builder.setPeriodic(pingFrequency * Constants.ONE_MIN_IN_MILLIS, 5 * Constants.ONE_MIN_IN_MILLIS);
-            builder.setRequiresBatteryNotLow(true);
-
-            if (Utils.hasPermission(context, "android.permission.RECEIVE_BOOT_COMPLETED")) {
-                builder.setPersisted(true);
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:575]: is this related to phone restart and is this handled?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java [hunk 3]
```diff
@@ -518,70 +478,84 @@ public Void call() {
...
+
+                // Todo - Earlier flex interval was 5 minutes but it was always defaulted to higher time? Possibly redundant to specify
+                PeriodicWorkRequest request = new PeriodicWorkRequest.Builder(CTPushAmpWorker.class, pingFrequency,
+                        TimeUnit.MINUTES)
+                        .setConstraints(constraints)
+                        .build();
+
+                String workName = existingWorkName.equals("") ? config.getAccountId() : existingWorkName;
+
+                workManager.enqueueUniquePeriodicWork(workName, ExistingPeriodicWorkPolicy.REPLACE, request);
```
[the remark this answers is not in the document]
Comment (darshanclevertap) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:535]: @CTLalit Yes, to support multiple instances. Multi-instance is where different instances of CleverTapAPI are invoked on the same app to send date to different accounts.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:1]: why not convert to kotlin?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/amp/CTPushAmpWorker.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import androidx.work.Worker
+import androidx.work.WorkerParameters
+import com.clevertap.android.sdk.CleverTapAPI
+import com.clevertap.android.sdk.Logger
+
+class CTPushAmpWorker(context: Context, params: WorkerParameters) : Worker(context, params) {
+
+    override fun doWork(): Result {
+        Logger.v("PushAmpWorker is awake")
+        CleverTapAPI.runJobWork(applicationContext)
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/amp/CTPushAmpWorker.kt:13]: do we need asynctask to start in this?

**Review by CTLalit — COMMENTED** [review]
Check comments.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java (hunk 2 above)
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:575]: WorkManager automatically handles this on it's own
https://stackoverflow.com/questions/50510559/androidx-workmanager-how-to-make-a-scheduled-job-persistent

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java (hunk 1 above)
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:483]: Done

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java [hunk 5]
```diff
@@ -518,70 +478,84 @@ public Void call() {
...
+            return;
         }
 
-        if (needsCreate) {
-            int jobid = config.getAccountId().hashCode();
-            JobInfo.Builder builder = new JobInfo.Builder(jobid, componentName);
-            builder.setRequiredNetworkType(JobInfo.NETWORK_TYPE_ANY);
-            builder.setRequiresCharging(false);
-
-            builder.setPeriodic(pingFrequency * Constants.ONE_MIN_IN_MILLIS, 5 * Constants.ONE_MIN_IN_MILLIS);
```
Comment (Anush-Shand) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:571]: @darshanclevertap @piyush-kukadiya We have 5 minutes specified as the flex interval for the job. Any specific reason and do we need the same for work manager?

Comment (darshanclevertap) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:571]: @Anush-Shand `pingFrequency` comes from the server under the key `pf` which is set as 240. So in production, the actual value is 240 mins or 4 hours. 5 mins is the fallback option in case the server does not send it.

Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:571]: @darshanclevertap pingfrequency is defaulted to 4 hours in the sdk as well. The second parameter (currently 5 minutes) specifies the flex interval i.e the window in the period (4 hours) in which the Job/Work can be completed
https://developer.android.com/develop/background-work/background-tasks/persistent/getting-started/define-work#flexible_run_intervals

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/amp/CTPushAmpWorker.kt (hunk 4 above)
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/amp/CTPushAmpWorker.kt:13]: This is not required as the worker is already running on a different thread. The db read is synchronised and the network requests are by default made on postAsyncSafely
Changing the same

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java (hunk 5 above)
Comment (darshanclevertap) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:571]: @Anush-Shand Ah, yes. So `setPeriodic` has interval and flex. If you do not provide flex, it will run anytime within the given interval time. We wanted it to be close to every 4 hours, so we set the flex as 5 mins. Hence the Job will run every 4 hours in the last 5 mins. This will make it truly periodic

Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:571]: @darshanclevertap 
Got it, we can add the same for work manager. But it reduces the chances of the work actually being completed for that period since there is only a window of 5 minutes to match the constraints

Comment (darshanclevertap) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:571]: @Anush-Shand I think we can do with that. Since it will work in the next run even if one interval is missed.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:534]: This has been handled. There is one periodic work request per app. The worker when invoked calls runPushAmpWork() for every available instance

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java [hunk 6]
```diff
@@ -443,145 +441,120 @@ private boolean alreadyHaveToken(String newToken, PushType pushType) {
...
+        int minute = now.get(Calendar.MINUTE);
 
-                if (lastTS == 0 || lastTS > System.currentTimeMillis() - 24 * 60 * 60 * 1000) {
-                    try {
-                        JSONObject eventObject = new JSONObject();
-                        eventObject.put("bk", 1);
-                        analyticsManager.sendPingEvent(eventObject);
+        Date currentTime = parseTimeToDate(hour + ":" + minute);
+        Date startTime = parseTimeToDate(Constants.DND_START);
+        Date endTime = parseTimeToDate(Constants.DND_STOP);
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:458]: We should not create new date object everytime.

**Review by CTLalit — COMMENTED** [review]
Comments added

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java (hunk 6 above)
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:458]: As in the startTime and the endTime can be Date constants itself rather than having DND_START and DND_STOP?

Comment (CTLalit) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:458]: we cannot have them as fixed constants i guess because we have to compute them for daily timestamps of DND hours. But the SDF object can be reused, you can explore if there are better ways to do this

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:1]: Can be done. The file is quite extensive though

Comment (CTLalit) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:1]: you can take a call for this and resolve.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/amp/CTPushAmpWorker.kt (hunk 4 above)
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/amp/CTPushAmpWorker.kt:13]: @piyush-kukadiya Can you please check this?
The Worker itself runs on a separate thread and hence postAsync is not required. 
The db read in runJobWork() is synchronised and the network requests are later by default made on postAsyncSafely

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:1]: Sure. I'll pick this up if I have the bandwidth before the release

File: clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java (hunk 6 above)
Comment (Anush-Shand) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/pushnotification/PushProviders.java:461]: This has been addressed

**Review by CTLalit — APPROVED** [review]
